awaiting nested AgentTask in a tool call within parent Agent's on_enter() stops working since v1.6.1

### Bug Description

I'm trying to migrate to LiveKit agents 1.6.1+ versions, and found some incompatibility between the newer versions and our previous evaluation/testing frameworks based on: https://docs.livekit.io/agents/start/testing/ . Specifically, the behavior of how to await nested AgentTask from a tool call within the parent Agent's `on_enter()` is changing and we would like your guidance.

I have a minimal reproducible script based on an earlier issue: https://github.com/livekit/agents/issues/5371#issuecomment-4203948109 (a related previous PR: https://github.com/livekit/agents/pull/5377) See below for the script and error info.


The following error is obtained with livekit-agents v1.6.4, but the same error appeared since v1.6.1.
```
ERROR    livekit.agents:log.py:22 Error in _llm_inference_task
Traceback (most recent call last):
  File "/Users/cj/Library/Caches/pypoetry/virtualenvs/agent-eval-2eDZ1Xtk-py3.11/lib/python3.11/site-packages/livekit/agents/utils/log.py", line 17, in async_fn_logs
    return await fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/cj/Library/Caches/pypoetry/virtualenvs/agent-eval-2eDZ1Xtk-py3.11/lib/python3.11/site-packages/opentelemetry/util/_decorator.py", line 71, in async_wrapper
    return await func(*args, **kwargs)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/cj/Library/Caches/pypoetry/virtualenvs/agent-eval-2eDZ1Xtk-py3.11/lib/python3.11/site-packages/livekit/agents/voice/generation.py", line 145, in _llm_inference_task
    async for chunk in llm_node:
  File "/Users/cj/Library/Caches/pypoetry/virtualenvs/agent-eval-2eDZ1Xtk-py3.11/lib/python3.11/site-packages/livekit/agents/voice/agent.py", line 476, in llm_node
    activity = agent._get_activity_or_raise()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/cj/Library/Caches/pypoetry/virtualenvs/agent-eval-2eDZ1Xtk-py3.11/lib/python3.11/site-packages/livekit/agents/voice/agent.py", line 415, in _get_activity_or_raise
    raise RuntimeError("no activity context found, the agent is not running")
RuntimeError: no activity context found, the agent is not running
WARNING  livekit.agents:agent_session.py:1463 session is closing, skipping start activity of agent_task1
WARNING  livekit.agents:generation.py:729 ToolError while executing tool: activity doesn't start for agent_task1, likely due to session closing
WARNING  livekit.agents:agent_activity.py:1474 attempting to schedule a new SpeechHandle, but the scheduling_task is not running, the speech will be cancelled
FAILED
```

Based on the pattern in livekit-agents’ own test for nested AgentTask on Github: https://github.com/livekit/agents/blob/main/tests/test_nested_agent_task.py , I modified the script accordingly to make it work with v1.6.1+. The change is to remove the `on_enter()` method from TestAgent.

### Expected Behavior

The original script should continue to work, or if that's an intended behavioral change (breaking existing tests), a clarification paragraph could be added to the official documentation at https://docs.livekit.io/agents/start/testing/ and https://docs.livekit.io/agents/logic/tasks/

### Reproduction Steps
The minimal reproducible script: https://github.com/livekit/agents/issues/5371#issuecomment-4203948109 (A related previous PR: https://github.com/livekit/agents/pull/5377)

```python
"""
LIVEKIT_EVALS_VERBOSE=1 pytest -s -o log_cli=true test_nested_task.py
Works for livekit-agents versions: 1.5.19rc1, 1.6.0, but not with 1.6.1

1.6.1 fails maybe due to this PR: https://github.com/livekit/agents/pull/6071

"""

import asyncio
# from typing import override

from livekit.agents import Agent, AgentSession, AgentTask, function_tool
from livekit.agents.llm.tool_context import NamedToolChoice
from livekit.plugins import openai
import pytest


class AgentTask1(AgentTask):
    def __init__(self) -> None:
        super().__init__(instructions="say hello")
        print('after init AgentTask1')

    # @override
    async def on_enter(self) -> None:
        print("AgentTask1 on_enter")
        tool = NamedToolChoice(type="function", function={"name": "run_agent_task_2"})
        await self.session.generate_reply(tool_choice=tool)

    @function_tool
    async def run_agent_task_2(self) -> None:
        print("run_agent_task_2")
        await AgentTask2()
        self.complete(None)


class AgentTask2(AgentTask):
    def __init__(self) -> None:
        super().__init__(instructions="say hello")

    # @override
    async def on_enter(self) -> None:
        print("AgentTask2 on_enter")
        tool = NamedToolChoice(type="function", function={"name": "run_agent_task_3"})
        await self.session.generate_reply(tool_choice=tool)

    @function_tool
    async def run_agent_task_3(self) -> None:
        print("run_agent_task_3")
        await AgentTask3()
        self.complete(None)


class AgentTask3(AgentTask):
    def __init__(self) -> None:
        super().__init__(instructions="say hello")

    # @override
    async def on_enter(self) -> None:
        print("AgentTask3 on_enter")
        await self.session.generate_reply(instructions="say hello")

    @function_tool
    async def goodbye(self) -> None:
        print("goodbye")
        self.complete(None)


class TestAgent(Agent):
    def __init__(self) -> None:
        super().__init__(instructions="say hello")

    # @override
    async def on_enter(self) -> None:
        print("\nTestAgent on_enter")
        tool = NamedToolChoice(type="function", function={"name": "run_agent_task_1"})
        await self.session.generate_reply(tool_choice=tool)

    @function_tool
    async def run_agent_task_1(self) -> None:
        print("run_agent_task_1")
        await AgentTask1()


@pytest.mark.asyncio
async def test_assistant_greeting() -> None:
    llm = openai.LLM(model="gpt-5.4-mini")
    async with AgentSession(llm=llm) as session:
        await session.start(TestAgent())
        print("user turn 1")
        await session.run(user_input="hi")
        print("user turn 2")
        await session.run(user_input="hi there")  # hangs here
        print("user turn 3")
        await session.run(user_input="sayonara")
```

Based on the pattern in livekit-agents’ own test for nested AgentTask on Github: https://github.com/livekit/agents/blob/main/tests/test_nested_agent_task.py , I modified the script accordingly to make it work with v1.6.1+. The change is to remove the `on_enter()` method from TestAgent.


### Operating System

macOS Sequoia

### Models Used

llm = openai.LLM(model="gpt-5.4-mini")

### Package Versions

```bash
livekit-agents == v1.6.1 or newer
```

### Session/Room/Call IDs

_No response_

### Proposed Solution

```python

```

### Additional Context

_No response_

### Screenshots and Recordings

_No response_

--- a/livekit-agents/livekit/agents/voice/agent.py
+++ b/livekit-agents/livekit/agents/voice/agent.py
@@ -873,6 +873,12 @@ class AgentTask(Agent, Generic[TaskResult_T]):
         # won't wait for tasks blocked on this handoff
         old_activity._add_drain_blocked_tasks(blocked_tasks)
 
+        # watch the blocked tasks so an active run won't complete mid-handoff
+        # (the parent speech may predate the run, e.g. created in on_enter)
+        if (run_state := session._global_run_state) and not run_state.done():
+            for task in blocked_tasks:
+                run_state._watch_handle(task)
+
         if (
             task_info.function_call
             and isinstance(old_activity.llm, RealtimeModel)

--- a/tests/test_nested_agent_task.py
+++ b/tests/test_nested_agent_task.py
@@ -74,6 +74,84 @@ async def test_nested_agent_task_no_deadlock():
         assert second_result is not None
 
 
+class SimpleTask(AgentTask):
+    """A task that needs a user turn to complete (user must trigger 'finish')."""
+
+    def __init__(self) -> None:
+        super().__init__(instructions="simple task")
+
+    async def on_enter(self) -> None:
+        self.session.generate_reply(instructions="task_greeting")
+
+    @function_tool
+    async def finish(self, ctx: RunContext) -> str:
+        """Called to complete the task."""
+        self.complete(None)
+        return "done"
+
+
+class EnterHandoffAgent(Agent):
+    """Agent whose on_enter reply calls a tool that awaits an AgentTask.
+
+    The on_enter speech predates any session.run(), so the run doesn't watch it.
+    """
+
+    def __init__(self) -> None:
+        super().__init__(instructions="root agent")
+
+    async def on_enter(self) -> None:
+        await self.session.generate_reply(instructions="enter_greeting")
+
+    @function_tool
+    async def start_task(self, ctx: RunContext) -> str:
+        """Transitions into SimpleTask."""
+        await SimpleTask()
+        return "task completed"
+
+
+@pytest.mark.asyncio
+async def test_handoff_from_pre_run_speech():
+    """A handoff triggered by a speech created before session.run() (e.g. in
+    on_enter) must keep the run alive until the new activity has started;
+    otherwise the next run() races the transition and gets rejected."""
+    llm = FakeLLM(
+        fake_responses=[
+            # on_enter generate_reply(instructions="enter_greeting") -> calls start_task;
+            # slow enough that run(user_input="hi") starts before the tool call lands
+            FakeLLMResponse(
+                input="enter_greeting",
+                content="",
+                ttft=1.0,
+                duration=1.0,
+                tool_calls=[FunctionToolCall(name="start_task", arguments="{}", call_id="call_1")],
+            ),
+            # user says "hi" while the handoff is in flight; this is the only
+            # speech the first run watches
+            FakeLLMResponse(input="hi", content="hello!", ttft=1.0, duration=2.0),
+            # SimpleTask on_enter greeting
+            FakeLLMResponse(input="task_greeting", content="hello from task", ttft=0, duration=0),
+            # user says "bye" -> LLM calls finish
+            FakeLLMResponse(
+                input="bye",
+                content="",
+                ttft=0,
+                duration=0,
+                tool_calls=[FunctionToolCall(name="finish", arguments="{}", call_id="call_2")],
+            ),
+            # after start_task tool output, LLM responds
+            FakeLLMResponse(input="task completed", content="all done", ttft=0, duration=0),
+        ]
+    )
+    async with AgentSession(llm=llm) as sess:
+        await sess.start(EnterHandoffAgent())
+
+        await asyncio.wait_for(sess.run(user_input="hi"), timeout=5.0)
+        assert isinstance(sess.current_agent, SimpleTask)
+
+        await asyncio.wait_for(sess.run(user_input="bye"), timeout=5.0)
+        assert isinstance(sess.current_agent, EnterHandoffAgent)
+
+
 def _build_fake_llm() -> FakeLLM:
     return FakeLLM(
         fake_responses=[

